### DISCUSSION TITLE
Who's using gqlgen in production?

### DISCUSSION THREAD

**Discussion by vektah** [99designs/gqlgen#1864] [Show and tell]
We are preparing a blog post and would like to get a feel for who is currently using gqlgen?

Might as well go first; 99designs has been running gqlgen for several months now.

**Comment by vvakame** [comment]
not yet in production... 😿

**Comment by ghost** [comment]
Using with the flutter graphql client.
Not in production yet but getting there

**Comment by JulienBreux** [comment]
Soon ~10 days

**Comment by jonlundy** [comment]
Running in production with one internal application for about 3 months now. User base 10-15.

**Comment by anurag** [comment]
[Render](https://render.com) is using it for everything in production and internal tooling. We love it.

**Comment by rahulmadduluri** [comment]
Is anyone using the subscriptions in production?

**Comment by jacklaaa89** [comment]
We are close to deploying at least 3-4 of our core services at Sky Bet using gqlgen.

**Comment by yamitzky** [comment]
We are using for our news app [newsdigest](https://app.newsdigest.jp/)'s backend API for two months!

For our use case, performance is very very very important, and gqlgen is awesome for that. :) I compared with other GraphQL libraries and choosed it.

**Comment by aaahrens** [comment]
@yamitzky how big is the performance jump in this versus gophers (a guesstimate is gucci). I know that the other libraries aren't that performant.

**Comment by codyleyhan** [comment]
[Neighborly](https://neighborly.com/) is starting to move one of our production applications to gqlgen

**Comment by marczahn** [comment]
After testing out gqglen and some other libraries we will use it in one of the next projects.

**Comment by adamgoose** [comment]
I use gqlgen almost daily. It is the foundation of how we build our GraphQL implementations, of which there are many. We write medical practice-management and automation software. gqlgen is deployed in linux containers, Windows desktop and service applications, as well as on Raspberry Pis!

Thank you so much for an excellent package! It has helped lay the foundation for most of our software development today!

**Comment by adamgoose** [comment]
> @rahulmadduluri : Is anyone using the subscriptions in production?

Yessir, we are! Our implementation is primitive, and our subscriptions are limited by it. Here's a simple example of some resolvers that might help you out. I haven't run this at all, but have copied some code from another project. No promises that it works as-is!

```golang
// Subscriptions stores active subscriptions. The index of the top-level map is either a Task ID, or the string "*". The second-level map stores an active subscription with a unique ID.
var Subscriptions = map[string]map[string]chan *models.Task
const SUB_ALL = "*"

func (r subscriptionResolver) TaskCompleted(ctx context.Context, id *string) (*models.Task, error) {
	subId := uuid.Must(uuid.NewV4()).String()
	c := make(chan *models.Charge, 1)

	if id == nil {
		id = &SUB_ALL
	}

	go func() {
		<-ctx.Done()
		delete(subs[id], subId)
	}()
	if subs[id] == nil {
		subs[id] = make(map[string]chan *models.Task, 0)
	}
	subs[id][subId] = c

	return c, nil
}

func (r mutationResolver) CompleteTask(ctx context.Context, id string) (*models.Task, error) {
	t := &models.Task{}
	// persist your task

        // Publish to filtered subscriptions
	for _, c := range Subscriptions[id] {
		c <- t
	}
        // Publish to non-filtered subscriptions
	for _, c := range Subscriptions[SUB_ALL] {
		c <- t
	}

	return t, nil
}
```

One thing we need to improve on is my eventing. With a proper event bus, subscription "filters" could be vastly more complex; not to mention all of the other benefits of an event-based system.

**Comment by MichaelMure** [comment]
[git-bug](https://github.com/MichaelMure/git-bug) is not fully ready ready but does use gqlgen to bind the web UI with the backend.

**Comment by icco** [comment]
I'm not a company, but I use it for the backend of my blog, writing.natwelch.com.

**Comment by frederikhors** [comment]
Anyone tested performances vs `gophers-graphql` library?

**Comment by vektah** [comment]
I think @lwc is working on some cross framework performance tests

**Comment by andrewmunro** [comment]
We are using gqlgen quite heavily at https://github.com/skybet for multiple projects, we were quite early adopters and have seen this library go through many changes in the last year. Amazing to see 99designs adopting and actively developing it.

We are using graphql to enable backend teams to build out APIs and automatically generate administrative front-ends to configure them.

Everything works great, our current pain points are around dataloaders and performance as our dataset has grown substantially since we started. Complexity has helped things, we are now able to rate limit people sending us stupid requests...adding things like pagination etc is next on our list.

**Comment by sdalezman** [comment]
[Intello](intello.io) is using gqlgen in almost all of our production services

**Comment by bmonkman** [comment]
We've been using it in production  at https://bench.co for a couple months now, moving more and more functionality over. It's on the edge of our infra, running in Kubernetes and fronted with an AWS ELB, then resolvers are making REST calls to our backend services to fetch data.  Right now we have about 50 of our models represented, and about 120 resolvers created. Really liking it so far, thanks for all the hard work, @vektah

**Comment by imiskolee** [comment]
`Bindo Labs Limited`  used it for an open model CRM platform.

https://bindopos.com

**Comment by wawoon** [comment]
I work for Welyco(https://welyco.com), which is a startup company runs smart restaurants in Japan.
In our restaurants, customers can order foods and make payments with their smartphone.

Our restaurant system is fully GraphQL API powered by gqlgen.
Now it's under development, but we will release it within 2 months.

**Comment by davcamer** [comment]
I'm using it at Digital Ocean for an internal hardware management system.

**Comment by LennyPenny** [comment]
Are there any open source projects using it? The examples don't really help with more complex apis

**Comment by MichaelMure** [comment]
@LennyPenny [git-bug](https://github.com/MichaelMure/git-bug/tree/master/graphql) does

**Comment by jonatasbaldin** [comment]
Hey there!

Just did an Exchange Rate API with gqlgen ✨ 

https://exchangeratesgraphql.deployeveryday.com/

**Comment by vvakame** [comment]
We started provides some page component by gqlgen ✨ 
https://techbookfest.org/

**Comment by sagikazarmark** [comment]
We at [Banzai Cloud](http://banzaicloud.com/) use gqlgen for the v2 API of our [Cloud Info](https://github.com/banzaicloud/cloudinfo) application and we love it.

**Comment by pkosiec** [comment]
We at [Kyma](https://kyma-project.io) use `gqlgen` as GraphQL API layer for Kubernetes resources. Kyma is an open-source project driven by SAP.

The component is called `console-backend-service` and the source code is located here: https://github.com/kyma-project/kyma/tree/master/components/console-backend-service

**Comment by stevenjack** [comment]
We use `gqlgen` in production at [Vidsy](https://vidsy.co) and ❤️it, @vektah and team congrats on such a well thought-out and executed alternative to the other `golang` implementations. Currently only for our client facing platform that's not publicly accessible but big plans for the future!

**Comment by DBL-Lee** [comment]
Check out [DLive](https://dlive.tv). We have migrated from Apollo Server and been running `gqlgen` as our API server for a few months and it works perfectly. Thank you for the amazing project @vektah !

**Comment by fwojciec** [comment]
There are many much more impressive projects listed here, but I wanted to add [Graal Literary Agency's website](https://graalagency.com) to the list.  The backend implemented using `gqlgen` powers a frontend app `Next.js` and an admin panel built using `React`. Thank you so much for this wonderful piece of software -- it's been a joy to build and work with!

**Comment by jakubknejzlik** [comment]
We are using it in production on few projects (although small ones). We also created library (still wip) which generates GraphQL with gqlgen resolvers using gorm which helps to create model services more easily. Checkout https://github.com/novacloudcz/graphql-orm if interested :)

**Comment by bowd** [comment]
The [monitoring](https://monitoring.aleth.io) product from the [Alethio](company.aleth.io) suite of products, which is part of [Consensys](consensys.net) is build using gqlgen 🚀 

Keep up the good work!

**Comment by jaredly** [comment]
We're using it at [Khan Academy](https://khanacademy.org).

**Comment by niondir** [comment]
I currently try to migrate our ~850 lines long schema from `github.com/graphql-go/graphql` to gqlgen which we use in production. Currently I'm positive that gqlgen will save me from writing a lot of boilerplate. 

I will keep this post updated if I have some progress :)

**Update**
After some work I got from ~7.600 self written lines to ~1700 lines of code (which is ~22.3%).
All tests are green and were quiet easy to migrate with a simple wrapper function that works for both frameworks:
`func DoRequest(req *GraphqlRequest) (map[string]interface{}, error)`

If someone is interested in this wrapper, let me know. But it's straight forward to write.

Plus IMHO using a schema driven approach helps with keeping Required fields, default values and the schema it self more clean.

**Comment by BitPhinix** [comment]
We at Barbra (https://barbra.io) use gqlgen + federation in production with 5+ Microservices. It's been a pleasure so far, thanks for this awesome project and keep up the good work.

**Comment by frederikhors** [comment]
@BitPhinix can I ask what typ of federation are you using?

**Comment by BitPhinix** [comment]
@frederikhors we are using Apollo federation with the WIP implementation from @marwan-at-work (+ a gateway written in nodejs) and we haven't encountered any issues so far. Since it's merged in the latest master, we will switch to it in the next release.

**Comment by and1truong** [comment]
@go1com is using gqlgen. Thank you for great tool.

**Comment by swkumar** [comment]
[Thumbtack](http://www.thumbtack.com) is using gqlgen in production for many of our APIs that power our native and web apps and are in the process of moving over several others.

**Comment by Crevil** [comment]
We @lunarway are using `gqlgen` for most of our backoffice solutions replacing OpenAPI and gRPC.

**Comment by seaskyways** [comment]
Soon in about 30 days. Still waiting for that Field directive info issue tho :D

**Comment by kauehmoreno** [comment]
We at globo.com use gqlgen on globoesporte.com

**Comment by bastianccm** [comment]
Hi, the https://github.com/i-love-flamingo/flamingo project uses gqlgen: https://github.com/i-love-flamingo/graphql
You can find some references build with Flamingo on https://flamingo.me

**Comment by aaad** [comment]
We will in two weeks. Thank you for your work!

**Comment by cayter** [comment]
Thanks to this awesome library for making GraphQL dev in golang a breeze. I'm currently using this in a [web framework](https://github.com/appist/appy) I'm working on that provides a smooth DX. 

For documentation details and GIF demo, please refer to [here](https://appist.gitbook.io/appy/). The web framework also provides auto-recompile upon changes for both HTTP server(gqlgen is built-in) and background job worker in `debug` mode.

**Comment by colinschoen** [comment]
We're using it at [Postmates](http://postmates.com/).

**Comment by NickBlow** [comment]
We're shortly about to put a GQLGen powered subscription server live at [Wakelet.com](https://wakelet.com/). Will probably migrate more to it as experience of working with has been very good. Found subscriptions much easier to work with than in Apollo.

I'm aiming to open source my Server Sent Events transport, but I can't promise anything yet...

**Comment by vinicio** [comment]
We have been using it at [Palco MP3](https://www.palcomp3.com.br/) APIs since 2018, dealing with an average of 200 req/s. 🤘 

Thank you!

**Comment by ravisastryk** [comment]
I use gqlgen for developing most of my backend services. It has become part of our foundational aspects in building unified APIs. We recently started using it at [Neustar](https://www.home.neustar/) and excited to see many companies using in production.

**Comment by ibrahimker** [comment]
We're using it at [Warung Pintar](https://warungpintar.co.id/) since late 2019 and were quite happy with it. Thanks for the awesome work

**Comment by vishnuitta** [comment]
we are researching to pick a go based graphql package for our opensource project.. and planning to use this. It would be good to know some great projects (including opensource ones) using this 
@vektah It would be helpful if you can point me to blog having references to the adapters of 99designs/gqlgen. Thanks.

**Comment by ghost** [comment]
> > @rahulmadduluri : Is anyone using the subscriptions in production?
> 
> Yessir, we are! Our implementation is primitive, and our subscriptions are limited by it. Here's a simple example of some resolvers that might help you out. I haven't run this at all, but have copied some code from another project. No promises that it works as-is!
> 
> ```go
> // Subscriptions stores active subscriptions. The index of the top-level map is either a Task ID, or the string "*". The second-level map stores an active subscription with a unique ID.
> var Subscriptions = map[string]map[string]chan *models.Task
> const SUB_ALL = "*"
> 
> func (r subscriptionResolver) TaskCompleted(ctx context.Context, id *string) (*models.Task, error) {
> 	subId := uuid.Must(uuid.NewV4()).String()
> 	c := make(chan *models.Charge, 1)
> 
> 	if id == nil {
> 		id = &SUB_ALL
> 	}
> 
> 	go func() {
> 		<-ctx.Done()
> 		delete(subs[id], subId)
> 	}()
> 	if subs[id] == nil {
> 		subs[id] = make(map[string]chan *models.Task, 0)
> 	}
> 	subs[id][subId] = c
> 
> 	return c, nil
> }
> 
> func (r mutationResolver) CompleteTask(ctx context.Context, id string) (*models.Task, error) {
> 	t := &models.Task{}
> 	// persist your task
> 
>         // Publish to filtered subscriptions
> 	for _, c := range Subscriptions[id] {
> 		c <- t
> 	}
>         // Publish to non-filtered subscriptions
> 	for _, c := range Subscriptions[SUB_ALL] {
> 		c <- t
> 	}
> 
> 	return t, nil
> }
> ```
> 
> One thing we need to improve on is my eventing. With a proper event bus, subscription "filters" could be vastly more complex; not to mention all of the other benefits of an event-based system.

Heya! Can I ask how you are implementing your `server.go` for this subscription? I just started to try gqlgen and haven't been able to get the websockets part to work, keep seeing errors in my graphql-playground. Would be very helpful if you could share a bit more! Thanks :)

**Comment by gonzaloserrano** [comment]
We at [Paack](https://paack.co/) are using it in prod since ~6 months ago.

**Comment by fredbi** [comment]
We have been using `gqlgen` in production at [OneConcern](https://www.oneconcern.com) for 2 years without issue.
Using subscriptions, and now adopting federation.

**Comment by jakubknejzlik** [comment]
@fredbi thanks for sharing. May I ask how are You handling the federation with subscriptions? Are You using federated gateway only for Query/Mutations and keep connecting to gqlgen service directly for subscriptions?

**Comment by fredbi** [comment]
@jakubknejzlik 

> thanks for sharing. May I ask how are You handling the federation with subscriptions? Are You using federated gateway only for Query/Mutations and keep connecting to gqlgen service directly for subscriptions?

Yes: client WS transport is routed directly to the WS backend. Regular queries and mutations are handled by the gateway

**Comment by mastercactapus** [comment]
We've been using it on [GoAlert](https://github.com/target/goalert) for almost 2 years.

**Comment by michaeloverton** [comment]
we have been using it at [the knot](https://www.theknot.com) for 6-8 months. we migrated to it from [graphql-go](https://github.com/graph-gophers/graphql-go) because of its better feature set and ease of use. thank you!

**Comment by diep-it-dn** [comment]
I'm building a Full-Stack webapp to make online tutorial in near future by using gqlgen, ent, Angular, docker then deploy to Heroku at here:
https://fullstack-go-angular.herokuapp.com

Register an account or use: test_create@test.com/abc@123 that has permissions to create users, permissionGroups but other permission is denied like update, delete...

**Comment by warent** [comment]
Happy to say I've recently rewritten and relaunched [my SaaS app](https://fontpeek.com) with gqlgen (paired with https://github.com/kyleconroy/sqlc), running happily and smoothly in production. Easily handling big spikes of 100+ requests per second on a decently small machine

These two libraries made writing a graphql server in golang a breeze!

**Comment by prabhuomkar** [comment]
We @torchexpo - [https://torchexpo.xyz](https://torchexpo.xyz) have moved our lazy Apollo GraphQL to gqlgen in 2 days.
So far, its giving a good performance with multiple data loaders. 

Would appreciate some recipes on pagination and focus on entity related pagination as well, e.g.:
```
query {
  publisher(id: "facebook") {
    models(limit: 10) {
      name
    }
  }
}
```

**Comment by khelif96** [comment]
Mongodb is using it in production to power our graphql api for our  [internal CI tool](https://github.com/evergreen-ci/evergreen).

**Comment by negz** [comment]
We used gqlgen to build https://github.com/upbound/xgql at @upbound, and are using it in production as of last week. xgql is a GraphQL API for @crossplaneio (and thus by extension @kubernetes). We have a somewhat novel model in that we deploy hundreds of xgqls as sort of 'sidecars' to Crossplane deployments in Kubernetes clusters, then use another layer of GraphQL (currently Apollo based) to stitch them all together into a SaaS app.

**Comment by txbrown** [comment]
> We @torchexpo - https://torchexpo.xyz have moved our lazy Apollo GraphQL to gqlgen in 2 days.
> So far, its giving a good performance with multiple data loaders.
> 
> Would appreciate some recipes on pagination and focus on entity related pagination as well, e.g.:
> 
> ```
> query {
>   publisher(id: "facebook") {
>     models(limit: 10) {
>       name
>     }
>   }
> }
> ```

> Mongodb is using it in production to power our graphql api for our [internal CI tool](https://github.com/evergreen-ci/evergreen).

I am maintaining this api starter project based on gqlgen that I use on a personal and client projects.
Its based off of this blog post and repo - https://dev.to/cmelgarejo/creating-an-opinionated-graphql-server-with-go-part-1-3g3l although I have added other bits to it and a clean architecture approach in the meantime.

A query would look like:

```
type Query {
  packs(
    id: ID
    filters: [QueryFilter]
    limit: Int = 50
    offset: Int = 0
    orderBy: String = "id"
    sortDirection: String = "ASC"
  ): [Pack!]!
}
```

You can see the parsing of the query params into sql [here](https://github.com/yonkolevel/gqlgen-api-starter/blob/main/internal/orm/tools.go)

**Comment by diep-it-dn** [comment]
> I'm building a Full-Stack webapp to make online tutorial in near future by using gqlgen, ent, Angular, docker then deploy to Heroku at here:
> https://fullstack-go-angular.herokuapp.com
> 
> Register an account or use: [test_create@test.com](mailto:test_create@test.com)/abc@123 that has permissions to create users, permissionGroups but other permission is denied like update, delete...

I’ve shared the repo at: https://github.com/diep-it-dn/fullstack-go-angular
Please view it for more details!

**Comment by StevenACoffman** [comment]
@vektah [Khan Academy](https://khanacademy.org/) currently has 91.1% of its requests being served via Go (and the majority of those uses gqlgen). For some sense of scale, last year millions of people spent 8.7 BILLION minutes actively learning on Khan Academy. 

I did a [high level talk about our transition from Python to Go](https://www.youtube.com/watch?v=je9bC3DZ6tg) and gqlgen is what made it all possible. We are very grateful to the contributors to this wonderful repository! ❤️ 

That said, if a maintainer gets a chance to [review our pull request for resolving requests for federation entities in parallel](https://github.com/99designs/gqlgen/pull/1285) we would be even more thrilled. 😏

**Comment by dilipkk-foyernet** [comment]
We are using at [Foyernet](https://foyernet.com/)!!!

**Comment by kaythiele** [comment]
We are using gqlgen at [www.everphone.com](https://www.everphone.com)

**Comment by mtibben** [comment]
> Happy to say I've recently rewritten and relaunched [my SaaS app](https://fontpeek.com) with gqlgen (paired with https://github.com/kyleconroy/sqlc), running happily and smoothly in production. Easily handling big spikes of 100+ requests per second on a decently small machine
> 
> These two libraries made writing a graphql server in golang a breeze!

Nice one @warent. I'd love to see a some kind of comparison between sqlc/gqlgen vs ent/gqlgen - both popular choices

**Comment by prabhuomkar** [comment]
Look at how [Skim](skimhq.tech) built Authentication and Authorization with gqlgen, a blog post by @janeadsouza:
https://janedsouza.xyz/blog/auth-with-gqlgen

**Comment by frederikhors** [comment]
@StevenACoffman can we pin this issue? Or should we use "Discussions" for this?

**Comment by StevenACoffman** [comment]
I think a discussion is probably good, right?

**Comment by edouard-claude** [comment]
I use it for everything on production and internal tools. I love that !!

**Comment by imrajdas** [comment]
[LitmusChaos](https://github.com/litmuschaos/litmus), A CNCF incubation project, We are using GQLGen (v0.11.3) for the last 2 years. Great tool ❤️

**Comment by ankitharma212** [comment]
We've faced similar challenges at [Ledger Labs](https://www.theledgerlabs.com/) when integrating GraphQL with Go, particularly with time handling. A solution that worked for us was defining custom scalar types in the GraphQL schema for time.Time. This allowed us to manage different time formats more effectively. Additionally, using middleware for time serialization and deserialization helped maintain consistency across our applications. Hope this helps!

**Comment by UnAfraid** [comment]
At work we use gqlgen for 6 years now, works great.
I also use it for a couple personal projects for example https://github.com/UnAfraid/wg-ui
